File: ffi/python/blsct/signature.py
import blsct
from .managed_obj import ManagedObj
from .public_key import PublicKey
from .scalar import Scalar
from typing import Any, Self, override

class Signature(ManagedObj):
  def __init__(self, obj):
    super().__init__(obj)

  @staticmethod
  def generate(priv_key: Scalar, msg: str) -> Self:
    sig = blsct.sign_message(priv_key.value(), msg)
    return Signature(sig)

  def verify(self, msg: str, pub_key: PublicKey) -> bool:
    return blsct.verify_msg_sig(pub_key.value(), msg, self.value())

  @override
  def value(self) -> Any:
    return blsct.cast_to_signature(self.obj)

  @override
  def default(self) -> Self:
    name = self.__class__.__name__
    raise NotImplementedError(f"{name}.default()")

File: ffi/ts/src/signature.ts
import {
  castToSignature,
  deserializeSignature,
  serializeSignature,
  signMessage,
  verifyMsgSig,
} from './blsct'

import { Scalar } from './scalar'
import { PublicKey } from './keys/publicKey'
import { ManagedObj } from './managedObj'

/** Represents the signature of a transaction.
 *
 * Examples:
 * ```ts
 * const { PublicKey, Scalar, Signature } = require('navio-blsct')
 * const sk = Scalar.random()
 * const pk = PublicKey.fromScalar(sk)
 * const sig = Signature.generate(sk, 'navio')
 * sig.verify(pk, 'navio')
 * const ser = sig.serialize()
 * const deser = Signature.deserialize(ser)
 * ser === deser.serialize() // true
 * ```
 */
export class Signature extends ManagedObj {
  constructor(obj: any) {
    super(obj)
  }

  /** Generates a new `Signature` instance.
   *
   * @param privKey - The private key used to sign the message.
   * @param msg - The message to be signed.
   * @return A new `Signature` instance containing the signature of the message.
   */
  static generate(
    privKey: Scalar,
    msg: string,
  ): Signature {
    const obj = signMessage(privKey.value(), msg)
    return new Signature(obj)
  }

  /** Verifies the signature against a public key and message.
   *
   * @param pubKey - The public key used to verify the signature.
   * @param msg - The message that was signed.
   * @return `true` if the signature is valid for the given public key and message, otherwise `false`.
   */
  verify(
    pubKey: PublicKey,
    msg: string,
  ): boolean {
    return verifyMsgSig(pubKey.value(), msg, this.value())
  }

  override value(): any {
    return castToSignature(this.obj)
  }

  override serialize(): string {
    return serializeSignature(this.value())
  }

  /** Serializes the signature to a hexadecimal string.
   *
   * @param hex - The hexadecimal string to serialize.
   * @returns A hexadecimal string representation of the signature.
   */
  static deserialize(
    this: new (obj: any) => Signature,
    hex: string
  ): Signature {
    return Signature._deserialize(hex, deserializeSignature)
  }
}

